### ISSUE TITLE
Aliasing data access in unions is sometimes not recognized

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#5362]
Build the following code

```c++
#include <stdio.h>
#include <stdlib.h>

struct pair
{
    mutable float x, y;
};

struct float2
{
    float2() {}
    float2(float x, float y):x(x), y(y) {}
    union
    {
        pair xy;
        struct
        {
            mutable float x, y;
        };
    };
    const float2 &operator=(const float2 &rhs) const
    {
        xy.x = rhs.xy.x;
        xy.y = rhs.xy.y;
        return *this;
    }
};

float2 min(const float2 &lhs, const float2 &rhs)
{
    return float2(lhs.x < rhs.x ? lhs.x : rhs.x, lhs.y < rhs.y ? lhs.y : rhs.y);
}

int main()
{
    float2 minimum(1.0f, 1.0f);
    float2 value;
    int one = (rand() == 0 && rand() == 10000) ? 10000 : 1; // Confuse LLVM so it does not know up front how many iterations there are (always exactly 1).
    for(int i = 0; i < one; i++)
    {
        value = float2(1.0f, 0.0f);
        printf("val: %f, %f\n", value.x, value.y);
        minimum = min(minimum, value);
        printf("min: %f, %f\n", minimum.x, minimum.y);
    }
}
```

with `em++ a.cpp -O2 -o a.js ; node a.js`, which prints

```
val: 1.000000, 0.000000
min: 0.000000, 0.000000
```

whereas it should print (and when omitting `-O2`, it does)

```
val: 1.000000, 0.000000
min: 1.000000, 0.000000
```

This issue is in upstream LLVM/Clang, since also building natively with Emscripten's LLVM with `clang++ a.cpp -O2 -o a ; ./a` gives the same incorrect output (omitting `-O2` again prints correct output)

**Comment by juj** [comment]
This issue is a regression in Emscripten, last working version is Emscripten 1.37.6, and the regression occurs in 1.37.7, when we updated to LLVM 4.0.

**Comment by juj** [comment]
Bisecting through upstream LLVM/Clang repository (testing with native `clang++` build and not `em++`), the following git commits are the last good working pair:
```
llvm:
commit ddd3bb716c63ca44bb7a1fa2d169cbce6a651052
Author: Dehao Chen <dehao@google.com>
Date:   Fri Dec 30 00:50:28 2016 +0000

clang:
commit 1008244ce7090576ea3a33e38017c893f30a97ab
Author: Kelvin Li <kkwli0@gmail.com>
Date:   Thu Dec 29 22:16:30 2016 +0000
 ```

but the next commit introduced to LLVM causes the breakage:

```
llvm:
commit 4f67d55a960b7694c8f2cb0aa0e2d9705b4bc0ee
Author: Michael Kuperstein <mkuper@google.com>
Date:   Fri Dec 30 01:03:17 2016 +0000

    [LICM] When promoting scalars, allow inserting stores to thread-local allocas.
    
    This is similar to the allocfn case - if an alloca is not captured, then it's
    necessarily thread-local.
    
    Differential Revision: https://reviews.llvm.org/D28170
    
    
    git-svn-id: https://llvm.org/svn/llvm-project/llvm/trunk@290738 91177308-0d34-0410-b5e6-96231b3b80d8

clang:
commit 1008244ce7090576ea3a33e38017c893f30a97ab
Author: Kelvin Li <kkwli0@gmail.com>
Date:   Thu Dec 29 22:16:30 2016 +0000
```

**Comment by juj** [comment]
However fortunately it looks like this breakage is already fixed in later upstream LLVM/Clang. The last breaking pair of git commits is

```
llvm:
commit 9693a6db68198e4994eb40365f3d1507e187a508
Author: James Molloy <james.molloy@arm.com>
Date:   Thu May 25 12:51:11 2017 +0000

clang:
commit 3f8116e6a2815b1d5f3491493938d0c63c9f42c9
Author: Egor Churaev <egor.churaev@intel.com>
Date:   Thu May 25 07:18:37 2017 +0000
```

whereas the next commit introduced to Clang (not LLVM) fixes the test case:

```
llvm:
commit 9693a6db68198e4994eb40365f3d1507e187a508
Author: James Molloy <james.molloy@arm.com>
Date:   Thu May 25 12:51:11 2017 +0000

clang:
commit 5e606f16eec95ead0f7907838ab5ba24318e7eda
Author: Krzysztof Parzyszek <kparzysz@codeaurora.org>
Date:   Thu May 25 12:55:47 2017 +0000

    [CodeGen] Pessimize aliasing for member unions (and may-alias) objects
    
    Use the TBAA info of the omnipotent char for these objects.
    
    Differential Revision: https://reviews.llvm.org/D33328
    
    
    git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@303851 91177308-0d34-0410-b5e6-96231b3b80d8
```

In particular it looks like that a commit in LLVM repository caused the breakage (by adding a new optimization it looks like), but a later commit in Clang repository fixes it (by making union aliasing detection stricter).

Currently our LLVM upstream comes from around ~March 15th or so, so we are in the breaking range. I wonder what would be the best path to get the fix, work towards merging to a newer upstream or to cherry pick?

**Comment by kripken** [comment]
Cherry-picking is the fastest thing. However, if someone has the time to do an LLVM update that would be [useful for Rust as well](https://internals.rust-lang.org/t/moving-webassembly-support-forward/5460/15).

**Comment by juj** [comment]
I have now tried to cherry pick the commits, but it looks like the one commit does not build clean, and trying to follow the trail of related changes leads to a messy looking graph of various commits that need to be pulled in to accommodate. I think the better approach here will be to merge us up to a newer version.

**Comment by juj** [comment]
This issue was fixed way back when we updated Emscripten to a newer LLVM 4.0.1/5.0svn in https://github.com/emscripten-core/emscripten-fastcomp/pull/191.

First fixed version is Emscripten 1.37.31 from 01/31/2018. The issue originated in LLVM 4.0 update that came about in at 1.37.7 from 3/15/2017.
